### DISCUSSION TITLE
Generate URL from routes

### DISCUSSION THREAD

**Discussion by jlaffaye** [tokio-rs/axum#1947] [Q&A]
Hello,

My goal is to re-use what the router already knows to generate an URL.

I found `TypedPath` in `axum-extra`, but this does not play nicely with nested routers.

I want to use nested router to make the prefix of the paths configurable with a command line flag.
So
```
/foo/bar
/hello/world
```

can be configured to be
```
/my-prefix/foo/bar
/my-prefix/hello/world
```

Is there a way to do what I want ?

**Comment by davidpdrsn** [comment]
`router.nest(path_from_commandline_args, other_router)`?

**Comment by jlaffaye** [comment]
Sorry, my question was not clear. Yes, I want to use `router.nest(path_from_commandline_args, other_router)` and then have some kind of API to get the URL for a route defined in `other_router` that returns the full URL `/my-prefix/foo/bar` (`path_from_commandline_args` being `/my-prefix`

**Reply by davidpdrsn** [reply]
axum intentionally doesn't support inspecting the routes in a `Router`. Implementing such a feature is a big effort and very closely tied with something like OpenAPI support, which we also don't have bandwidth to implement.

So you'll have to handle that yourself somehow.

**Reply by step21** [reply]
Is it possible to have something similar to provide the url in a template? I was asking at minijinja whether it can have url_for support, which is a templating function to provide urls with f.e. variable parts. https://github.com/mitsuhiko/minijinja/discussions/314#discussioncomment-7596215

**Reply by davidpdrsn** [reply]
I don't understand what you mean. Can you elaborate?

**Reply by step21** [reply]
Apologies. In template engines, in this case the jinja2 compatible minijinja, you can use some functions within template fields, for example to generate links based on the page you are on. The minijinja author said he thinks it is not possible with axum, and I just wanted to check.
You can find the doc for url_for here: https://flask.palletsprojects.com/en/3.0.x/api/#flask.url_for
I thought maybe you could use the https://docs.rs/axum/latest/axum/extract/struct.Query.html#impl-Clone-for-Query%3CT%3E
But I am still learning so maybe that does something different.

**Reply by davidpdrsn** [reply]
`Query` is for query params. Not the current page.

Do you mean extracting `axum::http::Uri`? That’ll give the URI of the request.

**Comment by jplatte** [comment]
axum does not have the concept of named routes the way Flask and actix-web do, so it doesn't have an equivalent to `url_for` either.

**Reply by step21** [reply]
ok thx

**Comment by punkstarman** [comment]
This makes it difficult to implement APIs that respect ReST's HATEOAS constraint.

**Comment by osmano807** [comment]
I'm using `axum_extra::routing::TypedPath` and using askama with a constructor function when using path parameters, like:

```rust
    #[derive(TypedPath, Deserialize, new)]
    #[typed_path("/")]
    pub struct Home;

    #[derive(TypedPath, Deserialize, new)]
    #[typed_path("/subpath/:layout")]
    pub struct Subpath {
        pub layout: Layout,
    }

Router::new()
            .typed_get(super::get::home)
            .typed_get(super::get::subpath)
```
```rust
      <a href="{{ crate::web::pages::routes::Home }}">Home</a>
<a
      href="{{ crate::web::pages::routes::Subpath::new(layout) }}"
      class="link menu-title"
    >
      Subpath {{ layout }}</span>
    </a>
```
